runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/commits/StreamingApplianceWorkCommitter.java

```diff
@@ -0,0 +1,183 @@
+    workCommitter.startCommitWorkers(ready);
+    return workCommitter;
+  }
+
+  @Override
+  public void commit(Commit commit) {
+    commitQueue.put(commit);
+  }
+
+  @Override
+  public long currentActiveCommitBytes() {
+    return activeCommitBytes.get();
+  }
+
+  @Override
+  public void stop() {
+    commitWorkers.shutdownNow();
+  }
+
+  @Override
+  public int parallelism() {
+    return numCommitWorkers;
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startCommitWorkers(CountDownLatch ready) {
+    for (int i = 0; i < numCommitWorkers; i++) {
+      commitWorkers.submit(
+          () -> {
+            while (true) {
+              try {
+                ready.await();
+                break;
+              } catch (InterruptedException ignore) {
+              }
+            }
+            commitLoop();
+          });
+    }
+  }
+
+  private void commitLoop() {
+    Map<ComputationState, Windmill.ComputationCommitWorkRequest.Builder> computationRequestMap =
+        new HashMap<>();
+    while (shouldCommitWork.get()) {
+      computationRequestMap.clear();
+      CommitWorkRequest.Builder commitRequestBuilder = CommitWorkRequest.newBuilder();
+      long commitBytes = 0;
+      // Block until we have a commit, then batch with additional commits.
+      Commit commit;
+      try {
+        commit = commitQueue.take();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        continue;
+      }
+      while (commit != null) {
+        ComputationState computationState = commit.computationState();
+        commit.work().setState(Work.State.COMMITTING);
+        Windmill.ComputationCommitWorkRequest.Builder computationRequestBuilder =
+            computationRequestMap.get(computationState);
+        if (computationRequestBuilder == null) {
+          computationRequestBuilder = commitRequestBuilder.addRequestsBuilder();
+          computationRequestBuilder.setComputationId(computationState.getComputationId());
+          computationRequestMap.put(computationState, computationRequestBuilder);
+        }
+        computationRequestBuilder.addRequests(commit.request());
+        // Send the request if we've exceeded the bytes or there is no more
+        // pending work.  commitBytes is a long, so this cannot overflow.
+        commitBytes += commit.getSize();
+        if (commitBytes >= TARGET_COMMIT_BUNDLE_BYTES) {
+          break;
+        }
+        commit = commitQueue.poll();
+      }
+      commitWork(commitRequestBuilder.build(), commitBytes);
+      completeWork(computationRequestMap);
+    }
+  }
+
+  private void commitWork(CommitWorkRequest commitRequest, long commitBytes) {
+    LOG.trace("Commit: {}", commitRequest);
+    activeCommitBytes.addAndGet(commitBytes);
+    commitWork.accept(commitRequest);
+    activeCommitBytes.addAndGet(-commitBytes);
+  }
+
+  private void completeWork(
```

Comment:
it seems like it woudl be better if this went through somethign like StreamingDataflowWorker.onStreamingCommitComplete handler instead of poking into the computationstate directly.

I think you could reuse the existing method, doesn't really seem streaming specific.

not using the computationState seems like it will make it easier to test as well.

Reply:
done